#<center> Mediciones directas, indirectas y fiteo de funciones</center>

In [ ]:
# Lista de paquetes que se importaron
import pandas as pd
import numpy as np
from sympy import *
import plotly.graph_objects as go

## Introducción de cinemática


En clase intentamos estimar el valor de la gravedad de una manera indirecta. Para esto, hicimos el experimento de dejar caer un borrador desde una cierta altura y modelamos su movimiento con el modelo físico que dice que los objetos que caen en caída libre sufren una aceleración constante.

\begin{equation}
\tag{1}
 A = g
\end{equation}

Sabiendo que la aceleración es constante, que la velocidad es su integral y que la velocidad inicial del borrador fue cero obtenemos:

\begin{equation}
\tag{2}
 V(t) = g.t
\end{equation}

Por último, utilizamos este razonamiento de que la integral de la velocidad es la posición para obtener

\begin{equation}
\tag{3}
y(t) = \frac{g.t^2}{2}+y_o
\end{equation}

Donde $y_o$ es la altura desde donde el borrador fue soltado.
Como podemos imaginarnos (o recordar) las magnitudes que pudimos medir fácilmente en el aula fueron la altura y el tiempo que tardó el borrador en ir desde la mano del asistente hasta el suelo. De la ecuación (3) podemos despejar el valor de la gravedad en función de estas cantidades:
\begin{equation}
\tag{4}
 g = \frac{-2.y_o}{t^2}
\end{equation}
El tiempo inicial (igual a cero) corresponde al instante en el cual el borrador deja de estar en contacto con la mano del asistente y el $t_c$ es el tiempo que transcurrió desde que se soltó el borrador hasta que los presentes vieron, o escucharon, al borrador impactar con el suelo. Teniendo en cuenta estos valores podemos decir que $y_o$ es la altura desde donde el borrador se soltó y que $y(t_c) = 0$.

Teniendo en claro todo esto 5 alumnos/as pasaron al frente y midieron con una cinta métrica la altura a la que se iba a soltar el borrador. Las medidas fueron las siguientes:



<a id="mediciones-clases"></a>
## Mediciones directas

In [ ]:
Yo = [2.7,2.68,2.68,2.59,2.71,2.62]

Como ven, las medidas fueron diferentes salvo dos que dieron

*   Elemento de lista
*   Elemento de lista

el mismo valor. Si alguien les preguntara desde que altura soltaron al borrador ¿que responderían? probablemente con el valor promedio

In [ ]:
y_promedio = sum(Yo)/len(Yo)
y_promedio

2.6633333333333336

Pero ésto hace que todas las medidas sean incorrectas porque ninguna es igual al promedio. Entonces para que no sea un problema dar medidas se escribe, aparte del valor promedio (valor mas probable), un margen o rango donde se van a encontrar el 68% de las mediciones. Es decir, que si hubieran medido la altura 100 personas, 68 de esas mediciones hubieran estado en ese rango. Este rango se llama "error", pero no porque sea una medicion errónea. Es algo así como el rango donde hay una cierta certeza de que el valor preciso esta "por ahí cerca". Este error resulta de la suma de todos los posibles factores que pueden acumular incertezas en la medición.
\begin{equation}
\tag{5}
 ΔE = \sqrt{ \sigma^2_{est}+\sigma^2_{nom}+...}
\end{equation}
Muchos de estos factores no se conocen, como por ejemplo, el angulo en el que estaba la cinta metrica cuando midió un alumno, o si el brazo del asistente no se movió, si el alumno tenia los lentes empañados e invento un número, etc. Para todas esas fuentes de incertezas se recurre al método de realizar muchas mediciones. Con todas esas mediciones se obtiene un rango estadístico de estas fuentes de errores. Este rango estadístico se le suele llamar *desviación estandar* o *sigma estadístico*.
\begin{equation}
\tag{6}
 \sigma_{est} = \sqrt{\frac{∑(\overline{y}-y_i)^2}{N}}
\end{equation}

En la ecuacion (5) hay otro sigma, el *sigma nominal*. Éste representa el valor mínimo que puede medir el instrumento utilizado. Es decir, *la precisión del instrumento*. En la mayoria de las cintas métricas la precision es de 1 cm, o dicho en el sistema MKS (Metros-Kilogramo-Segundos)  0.01 mt.
Calculemos todo esto para la altura que midieron los alumnos.

In [ ]:
for y in Yo:
  terminos = (y_promedio - y)**2
sigma_est = (terminos/len(Yo))**0.5
print(f'Sigma estadístico = {sigma_est}')

Sigma estadístico = 0.01769075925343411


Ahora para el cálculo total del error hay que considerar el sigma nominal de 0.01 mt.

In [ ]:
error = (0.01**2+sigma_est**2)**0.5
print(f'Error = {error}')

Error = 0.020321490175746575


El error siempre se va a expresar con una sola cifra significativa. Esto es porque a partir de esta cifra no podemos tener certezas sobre los valores. Entender ésto suele ser difícil y explicarlo un poco más. Les recomiendo leer el material bibliográfico de la cátedra para ahondar en el tema.
En definitiva lo que se pide es que el error tenga solo un dígito diferente a cero. No importa de qué lado de la coma esté ese dígito. Para hacer ésto vi esta solución <a>https://stackoverflow.com/questions/3410976/how-to-round-a-number-to-significant-figures-in-python</a>. Entonces el error queda:

In [ ]:
error_1cifra_sign = float('%.1g' % error)
print(f'error = {error_1cifra_sign}')


error = 0.02


Por ultimo hay que redondear el valor más probable, aquel que calculamos con el promedio, para que tenga la misma precisión que el error. Es decir, hasta el segundo dígito en este caso particular. De esta manera queda expresada la altura

\begin{equation}
\tag{7}
 y_o = [2.66 \pm 0.02] mts
\end{equation}

El estilo de mostrar valores tiene que ser respetado siempre. Al momento de calcular errores se va a revisar la cantidad de cifras significativas del error, el redondeo del valor más probable y que el valor final esté entre corchetes y con las unidades fuera de los mismos.

Luego de calcular la altura desde donde se iba a lanzar el borrador se midió el tiempo que tardó desde la mano del asistente hasta el suelo. Estos tiempos que los alumnos midieron fueron:

In [ ]:
t_list = [0.73,0.69,0.78,0.64,0.6,0.86,0.75,0.65,0.67,0.61,0.55,0.54,0.48,0.74,0.52,0.74,0.41,0.66,0.71,0.41,0.55,0.52,0.86,0.49,0.65,0.72,0.56,0.57,0.30,0.54,0.62,0.73,0.93]


Para calcular el promedio y la desviación estandar vamos a utilizar la libreria numpy y los metodos *mean* y *std*

In [ ]:
import numpy as np

In [ ]:
t_prom = np.mean(t_list)
t_desv_est = np.std(t_list)
print(f'Tiempo promedio = {t_prom} \nDesviacion estandart del tiempo = {t_desv_est}')

Tiempo promedio = 0.6296969696969696 
Desviacion estandart del tiempo = 0.1358806260519571


Cuidado! la desviación estandar del tiempo no es el error del tiempo (Ver anexo de histograma). Hay una fuente de error que falta considerar, el *tiempo de reacción*. Este tiempo es lo que tardamos en responder a un estimulo, en este caso al ruido del borrador al impactar con el suelo. Este tiempo suele estar ente los 0.1 y 0.2 segundos dependiendo de las personas. Ustedes pueden medirse ese tiempo con una regla (sí, con una regla, a buscar cómo) y compararlo con el valor que les da en esta pagina: https://reactiontest.io/. Sinceramente preferiría que reproduzcan esa app de manera local en un script de python. Supongo que ChatGPT les va a poder hacer el código fácilmente.
Volviendo al error del tiempo se calcularía de la siguiente manera:

\begin{equation}
\tag{8}
 ΔE_{tiempo} = \sqrt{ \sigma^2_{est}+\sigma^2_{reac}}
\end{equation}

In [ ]:
t_reac = 0.15
error_tiempo = np.sqrt(t_desv_est**2 + t_reac**2)
error_tiempo

0.20239452694248383

Entonces el tiempo de caida luego de redondear el error y el valor mas probable queda:
\begin{equation}
\tag{9}
 t_c = [0.6 \pm 0.2] segs
\end{equation}

## Mediciones indirectas
Ya estamos en condiciones de calcular el valor de gravedad. Miremos la ecuacion (4) Si a esta ecuación la evaluamos con los valores más probables de $y_o$ y de $t_c$ nos va a dar el valor mas probable de g
\begin{equation}
\tag{10}
 \overline{g} = \frac{-2.\overline{y_o}}{\overline{t}^2}
\end{equation}


In [ ]:
g_mas_prob = -2*y_promedio/(t_prom**2)
print(f'g mas probable = {g_mas_prob}')

g mas probable = -13.433596937901164


El valor que nos da este resultado está lejos de ser el mejor. Pero recuerden como midieron el tiempo. El error representa un 33% de la medicion. Esto quiere decir que la medición no fue para nada precisa. Lo único que podría salvar esta medición es el margen de error. Para calcularlo se recurre a la siguiente ecuación:

\begin{equation}
\tag{11}
 \Delta E = \sqrt{(\frac{\partial g}{\partial y_o}(\overline{y_o},\overline{t_c}))^2.(\Delta y_o)^2+(\frac{\partial g}{\partial t_c}(\overline{y_o},\overline{t_c}))^2.(\Delta t_c)^2}
\end{equation}

Para calcular las derivadas las pueden hacer a mano ya que son muy sencillas, o pueden aprovechar y aprender a utilizar el paquete sympy.

In [ ]:
from sympy import *

t = Symbol('t')
y = Symbol('y')

g = -2*y/(t**2)

g_t = g.diff(t)
g_y = g.diff(y)

print(f'Derivada de g respecto de t -> {g_t}')
print(f'Derivada de g respecto de y -> {g_y}')

Derivada de g respecto de t -> 4*y/t**3
Derivada de g respecto de y -> -2/t**2


Ahora se evalúan estas derivadas con los valores mas probables de $y_o$ y de $t_c$

In [ ]:
y_mas_prob = 2.67
t_mas_prob = 0.6

error_y = 0.02
error_t = 0.2

deriv_y = 4**y_mas_prob/(t_mas_prob**3)
deriv_t = -2/(t_mas_prob**2)

error_g = np.sqrt( (deriv_y)**2 * (error_y)**2+ (deriv_t)**2 * (error_t)**2)
print(f'El error de g = {error_g}')

El error de g = 3.911520466994073


Redondeando el error y el valor más probable llegamos a que el valor de la gravedad es:

\begin{equation}
\tag{12}
 g = [13 \pm 4] \frac{m}{seg^2}
\end{equation}

## Anexo de histograma

Vamos a analizar qué es el sigma estadístico. Tomemos por ejemplo los valores de tiempo que midieron los alumnos. Vamos a contar cuántos tomaron tiempos entre 0.3 y 0.4 segundos. Luego cuántos de 0.4 a 0.5 y asi hasta un segundo. Para eso utlizamos el método de numpy *histogram*.

In [ ]:
frec,bin_edges = np.histogram(t_list,bins = 7,range=(0.3,1))
print(f'Cantidad de alumnos: {frec}')
print(f'Bordes de bines: {bin_edges}')

Cantidad de alumnos: [1 4 8 9 8 2 1]
Bordes de bines: [0.3 0.4 0.5 0.6 0.7 0.8 0.9 1. ]


La idea es hacer un histograma. Un histograma es un grafico de barra que nos da la informacion que describimos arriba. Los bordes de estas barras es lo que devuelve el metodo np.histogram, asi que para graficar este histograma utilizando plotly, vamos a convertir los bordes de los bines (barras) a un solo valor que sea el punto medio.

In [ ]:
import plotly.graph_objects as go

In [ ]:
ancho_de_bin = (bin_edges[1] - bin_edges[0])/2
bins = bin_edges[:-1] + ancho_de_bin

fig = go.Figure()
fig.add_traces(go.Bar(x = bins,y = frec))

Como no me gusta como queda la grafica en cuanto a estetica les dejamos una funcion para que puedan jugar y dejar las graficas como les guste mas:

In [ ]:
def estetica(figura,x_name = '',y_name = '',title = '',w = 900,h = 450):
    figura.update_xaxes(showgrid=True,showline=True, linewidth=2, linecolor='black', mirror=True,title = x_name)#'cccccccccccccccc')
    figura.update_yaxes(showgrid=True,showline=True, linewidth=2, linecolor='black', mirror=True,title =y_name)


    figura.update_layout(
        margin=dict(l=20, r=20, t=40, b=20),
        plot_bgcolor="white",
        template="plotly_white",
        title=title,
        autosize=False,
        width=w,
        height=h,
        #yaxis = dict(range=[0,1]),
        #xaxis = dict(range=[0,1]),
        title_font = dict(size =20),
        legend=dict(orientation = 'v',
                    font = dict(size = 18),
                     yanchor="top",
                     y=0.95,
                     xanchor="right",
                     x=1,
                     bgcolor='rgba(255,255,255,0)',
                   )
    )

In [ ]:
fig = go.Figure()
fig.add_traces(go.Bar(x = bins,y = frec))
estetica(fig,title='Histograma de tiempos de caida',x_name = 'Tiempo (segs)',y_name = 'Cantidad de alumnos',h = 450,w = 900)
fig.show()

Ahora voy a graficar el mismo histograma pero dividiendo la altura de los bines por la cantidad de alumnos totales. Además voy a agregar tres líneas verticales: una para indicar el valor promedio de tiempo y otras dos que representen el tiempo más probable más un sigma estadístico y menos un sigma estadistico:

In [ ]:
fig = go.Figure()
fig.add_traces(go.Bar(x = bins,y = frec))
fig.add_vline(x=t_prom)
fig.add_vline(x=t_prom + t_desv_est,line_color = 'red')
fig.add_vline(x=t_prom - t_desv_est,line_color = 'red')
estetica(fig,title='Histograma de tiempos de caida',x_name = 'Tiempo (segs)',y_name = 'Cantidad de alumnos',h = 450,w = 900)

fig.show()


Como pueden ver, entre las lineas rojas se encuentran la mayoria de los valores, algo asi como el 68% de ellos. Viendo la forma del histograma podemos suponer que si sumamos muchos más alumnos vamos a tener un histograma similar, pero con las barras, o bines, más angostos. Como no tenemos más alumnos vamos a tener que recurrir a paquetes de python que dado el valor más probable y el valor de sigma simule valores que podrían llegar a medir otros alumnos.

In [ ]:
mu = t_prom
sigma = t_desv_est

#cantidad de alumnos simulados:
num_alumn = 10000
# data serían los tiempos que medirían los num_alumn simulados
data = np.random.normal(mu, sigma,num_alumn) # promedio, desv stand, cantidad de muestras


Vamos a repetir el procedimiento que hicimos para las mediciones de los alumnos reales, pero ahora suponiendo que tenemos más cantidad de alumnos/as. Como tenemos más alumnes podemos usar más bines en el histograma y más angostos.

In [ ]:
new_frec,new_bin_edges = np.histogram(data,bins = int(num_alumn/50))

new_ancho_de_bin = (new_bin_edges[1] - new_bin_edges[0])/2
new_bins = new_bin_edges[:-1] + new_ancho_de_bin

fig = go.Figure()
fig.add_traces(go.Bar(x = new_bins,y = new_frec/new_frec.sum()))
fig.add_vline(x=t_prom)
fig.add_vline(x=t_prom + t_desv_est,line_color = 'red')
fig.add_vline(x=t_prom - t_desv_est,line_color = 'red')
estetica(fig,title='Histograma de tiempos de caida',x_name = 'Tiempo (segs)',y_name = 'Cantidad de alumnos',h = 450,w = 900)

fig.show()


Para los que hicieron algún curso de estadística ya se darán cuenta que existe una función que puede describir la forma de este histograma. Esta función se conoce como *campana de Gauss* y describe la *distribucion normal* de densidad de probabilidad alrededor de un valor más probable.(ver https://en.wikipedia.org/wiki/Normal_distribution). Vamos a ver un método de python que devuelve los parámetros de esta distribución que mejor se ajustan a este histograma.[texto del enlace](https://)

In [ ]:
from scipy.optimize import curve_fit

def gauss(x,m,s):
    #La función PDF que está en la página https://en.wikipedia.org/wiki/Normal_distribution
    return (1/(s*(2*np.pi)**0.5))*np.exp(-0.5*((x-m)/s)**2)

In [ ]:
#la desindad de probabilidad se calcula como la frecuencia sobre la cantidad de mediciones por el ancho del bin
bin_widths = np.diff(new_bin_edges)
dens_prob = new_frec / (num_alumn * bin_widths)

popt, pcov = curve_fit(gauss, xdata = new_bins, ydata = dens_prob)
print('Parámetros optimos:',popt) # estos son los Parametros OPTimos. Es un array con los parametros óptimos en el orden en el que están en la función. En este caso primero "m" y luego "s"
print('Parámetros originales:', mu,sigma)

#Para obtener los errores de estos parámetros:
errs = np.sqrt(np.diag(pcov))
print('Errores de los parámetros de ajuste:',errs)

print('-------------------')
print(f'Valor mas probable ajustado t_prom = [{round(popt[0],3)} \u00B1 {round(errs[0],3)}] segs')
print(f'Valor mas probable ajustado sigma_t = [{round(popt[1],3)} \u00B1 {round(errs[1],3)}] segs')

Parámetros optimos: [0.62784939 0.13465629]
Parámetros originales: 0.6296969696969696 0.1358806260519571
Errores de los parámetros de ajuste: [0.00143903 0.00117498]
-------------------
Valor mas probable ajustado t_prom = [0.628 ± 0.001] segs
Valor mas probable ajustado sigma_t = [0.135 ± 0.001] segs


Presten atención a que el error de los parametros está en la tercer cifra decimal, justo donde el valor de la distribución y del ajuste empiezan a diferir. Para ser más claros, debajo graficamos la densidad de probabilidad y el histograma de los alumnos que generamos sinteticamente.

In [ ]:
fig = go.Figure()
fig.add_traces(go.Bar(x = new_bins,
                      y = dens_prob,
                      name = 'Histograma'))
fig.add_traces(go.Scatter(x = new_bins,
                          y = gauss(new_bins,popt[0],popt[1]),
                          name = 'Ajuste'))

Para volver a los datos reales, veamos como esto se puede hacer sin necesidad de crear mediciones sinteticas. Con solo los alumnos presentes se puede lograr el mismo ajuste. Aqui debajo representamos el histograma de densidad de probabilidad con el ajuste hecho anteriormente.

In [ ]:
fig = go.Figure()
fig.add_traces(go.Bar(x = bins,
                      y = frec/(len(t_list)*np.diff(bin_edges)),
                      name = 'Histograma'))
fig.add_traces(go.Scatter(x = new_bins,
                          y = gauss(new_bins,popt[0],popt[1]),
                          name = 'Ajuste'))

## Ajuste de funciones
Vamos a suponer que ustedes tienen la posicion de una piedra que tiran con una honda y la trackearon. Medir algo implica que cada medicion va a tener un error asociado. Este error se va a ver como "ruido". La siguiente entrada de codigo va a ser para simular datos medidos. Ustedes prueben con datos que ustedes hayan medido.

In [ ]:
# Defino condiciones iniciales de posicion y velocidad
xo = 0
yo = 10
vox = 1
voy = 20.3

t = [0]
x = [xo]
y = [yo]
while y[-1] >= 0:
  t += [t[-1] + 0.03]
  x += [vox*t[-1] + xo]
  y += [-9.8*(t[-1]**2)/2 + voy*t[-1] + yo]

# invento un ruido que le sumo a las posiciones
mu_ruido = 0 #genero un ruido centrado en cero
sigma_x = 0.01 # La desviacion del ruido respecto del cero
sigma_y = 0.01 # La desviacion del ruido respecto del cero

ruido_x = np.random.normal(mu_ruido, sigma_x, len(t))
ruido_y = np.random.normal(mu_ruido, sigma_y, len(t))

#Sumo el ruido a X y a Y
x = x + ruido_x
y = y + ruido_y

In [ ]:
fig = go.Figure()
fig.add_traces(go.Scatter(x = x, y = y, mode = 'markers'))
estetica(fig, title = 'Trayectoria (X en funcion de Y)', x_name = 'X', y_name = 'Y')
fig.show()

Veamos ahora como queda la velocidad en Y. Deberia quedar una funcion lineal como la siguiente:

\begin{equation}
\tag{13}
 V_y(t) = g.t +v_o^y
\end{equation}

Fijense que la pendiente es la gravedad y el termino independiente es la componente en y de la velocidad inicial.
El objetivo que vamos a perseguir va a ser obtener el valor de la gravedad a partir de datos medidos de un tiro oblicuo. Para esto encesitamos la posicion en Y, calcular su velocidad y ajustar con una recta con el metodo curve_fit. Primero grafiquemos $y(t)$

In [ ]:
fig = go.Figure()
fig.add_traces(go.Scatter(x = t,y = y, mode='markers'))
estetica(fig,title='Y en funcion del tiempo')
fig.show()

Ahora calculemos la velocidad. Para esto vamos a usar la libreria pandas asi aprovechamos metodos que nos pueden simplificar los calculos.

In [ ]:
import pandas as pd

In [ ]:
df = pd.DataFrame()
df['t'] = t
df['x'] = x
df['y'] = y
df['vel_x'] = df.x.diff()/df.t.diff()
df['vel_y'] = df.y.diff()/df.t.diff()
df

t         x          y     vel_x      vel_y
0    0.00 -0.001603   9.991384       NaN        NaN
1    0.03  0.027790  10.593248  0.979783  20.062106
2    0.06  0.074391  11.195651  1.553347  20.080124
3    0.09  0.087981  11.776673  0.453002  19.367386
4    0.12  0.105127  12.369191  0.571528  19.750589
..    ...       ...        ...       ...        ...
149  4.47  4.451914   2.822987  0.485507 -23.635804
150  4.50  4.508214   2.119092  1.876659 -23.463162
151  4.53  4.515145   1.395788  0.231030 -24.110125
152  4.56  4.547870   0.684472  1.090819 -23.710558
153  4.59  4.603936  -0.063669  1.868865 -24.938011

[154 rows x 5 columns]

In [ ]:
fig = go.Figure()
fig.add_traces(go.Scatter(x = df.t, y = df.vel_y, mode = 'markers'))

Como se puede ver en la grafica la velocidad es claramente una recta pero tiene valores que no podrian ubicarse encima de la recta adecuada. De todas formas vamos a poder ajustar estos puntos a la recta que mejor se ajuste.

In [ ]:
def velocity(t,g,vo):
  return(g*t+vo)

df = df.dropna().copy() # Elimino las filas que tienen NaNs

popt, pcov = curve_fit(velocity, xdata = df.t, ydata = df.vel_y)
errs = np.sqrt(np.diag(pcov))
print(popt,errs)

[-9.80444353 20.45762116] [0.02904181 0.07733906]


De esta manera obtuvimos que la gravedad es:
\begin{equation}
\tag{14}
 g = [-9.80 \pm 0.03] \frac{m}{seg^2}
\end{equation}

Por ultimo vamos a graficar la recta sobre los datos.

In [ ]:
fig = go.Figure()
fig.add_traces(go.Scatter(x = df.t, y = df.vel_y, mode = 'markers'))
fig.add_traces(go.Scatter(x = df.t, y = velocity(df.t,popt[0],popt[1])))

## Ejercicios.
Estos ejercicios tienen que mostrarselos a sus ayudantes a cargo.

1. Investigar cómo tomarse el tiempo de reaccion con una regla. Obtener el tiempo de reacción de cada uno midiéndoselo unas 10 veces cada uno. El valor tiene que estar expresado con su respectivo error.

2. Arrojar algo hacia arriba y tomando el tiempo de vuelo entre varios calcular con su error el valor de la altura maxima y la velocidad inicial suponiendo la gravedad como $[9.8 \pm 0.2] \frac{m}{seg^2}$  

3. Utilizar teoria de errores para mediciones directas, mediciones indirectas y ajuste de funciones en todo lo que se pueda en su proyecto.